### PR TITLE
Refactor price service and enable stale price lifetime

### PR DESCRIPTION
**Pull request by bbalser**
Refactored price service to have 2 modes
1.  If token key is configured, then it will get new price from solana and output on configured cadence. It will also absorb downtimes and use the stale price up to the configured duration
2. If no key is configured, but a default value is, it will output the default price on the configured cadence
3. If no key or default price is configured, the price generator will stop with a value ` Ok(())` for that token.

### CODE REVIEW

**Comment by maplant** [comment]
Instead of using anyhow, since this an API we should have an thiserror enum error type for this crate

File: price/src/price_generator.rs [hunk 1]
```diff
@@ -11,13 +11,13 @@ use tokio::time;
 
 #[derive(Debug, Clone, Serialize)]
 pub struct Price {
-    pub timestamp: i64,
+    pub timestamp: DateTime<Utc>,
```
Comment (maplant) [inline price/src/price_generator.rs:14]: Is it necessary for all of these fields to be public?

File: price/src/price_generator.rs [hunk 2]
```diff
@@ -11,13 +11,13 @@ use tokio::time;
...
     pub price: u64,
     pub token_type: BlockchainTokenTypeV1,
 }
 
 impl Price {
-    pub fn new(timestamp: i64, price: u64, token_type: BlockchainTokenTypeV1) -> Self {
+    pub fn new(timestamp: DateTime<Utc>, price: u64, token_type: BlockchainTokenTypeV1) -> Self {
```
Comment (maplant) [inline price/src/price_generator.rs:20]: I'd recommend making this new function `pub(crate)`. I don't think there's ever a need to construct this outside of this crate

**Review by jeffgrunewald — APPROVED** [review]

File: price/src/price_generator.rs (hunk 2 above)
Comment (bbalser) [reply price/src/price_generator.rs]: I just removed pub and everything compiled 29f472a

File: price/src/price_generator.rs [hunk 3]
```diff
@@ -59,148 +67,153 @@ impl TryFrom<PriceReportV1> for Price {
...
-                    Metrics::update(
-                        "price_stale_counter".to_string(),
-                        token_type,
-                        old_price.price as f64,
-                    );
-                    tracing::warn!("rpc failure: {err:?}, reusing old price {:?}", old_price);
-                    Some(old_price)
+            Err(err) => {
+                tracing::error!(
+                    "error is retrieving new price for {:?}: {err:?}",
```
Comment (vihu) [inline price/src/price_generator.rs:175]:
```suggestion
                    "error in retrieving new price for {:?}: {err:?}",
```

**Pull request merged**